developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,349 @@
+connection IPC messages, involving connection points, from the guest before
+sending them to the DSP.
+
+3.2. Streaming
+--------------
+
+Guest audio streaming is mostly transparent for the host. The host audio
+subsystem doesn't get involved with most guest streaming or kcontrol operations
+because the virtualisation is applied at the DSP level, not at the audio level.
+All those operations are completely processed on the guest and only resulting
+DSP IPCs are forwarded to the host. The only exceptions are beginning and ending
+of the streaming, at which times the ASoC subsystem on the host has to be
+informed, that a certain pipeline is becoming active or inactive respectively.
+
+The SOF vhost driver uses the STREAM_PCM_PARAMS IPC message from the guest to
+allocate an ALSA PCM runtime object and configure the HDA hardware. If zero-copy
+is not used and audio data is transferred via VirtQueues, DMA buffers and also
+allocated.
+
+Upon reception of the STREAM_TRIG_START IPC message the vhost driver activates
+the associated pipeline and updates the DPCM routing information on the host.
+
+As mentioned above, it is our goal to implement data zero-copy. As long as
+this isn't the case, the guest VirtIO driver implements .block_write() and
+.block_read() SOF DSP operation methods for data streaming via the data
+VirtQueue.
+
+The host receives requests on the data VirtQueue and copies data between PCM
+runtime buffers and the queue and responds either with a status or with a data
+buffer.
+
+Similarly to the data, streaming buffer position updates are currently also
+transferred via a dedicated VirtQueue. Both the host and the guest are
+configured to use IPC messages for position updates. Buffers for VirtIO
+VirtQueues are always provided by guests, therefore there should always be a
+position update buffer waiting on the host side to send a position update
+message as soon as one arrives from the DSP.
+
+Upon reception of the STREAM_TRIG_STOP IPC message the vhost driver updates
+the DPCM routing information and deactivates the virtual PCM pipeline, described
+in the `2.2. Topology`_ section above, on the host. This operation requires
+particular care, because in the present state, just calling
+
+|	soc_dpcm_runtime_update();
+|	snd_soc_runtime_deactivate();
+
+doesn't deactivate the pipeline and therefore doesn't allow runtime PM to
+suspend the interface, which then breaks following activation attempts. The
+current SOF VirtIO implementation contains a fix for that, which has to be
+upstreamed along with the rest of the ALSA core virtualisation modifications
+and extensions.
+
+3.3. Deactivation
+-----------------
+
+When the guest isn't actively using audio, we want to allow the host to
+runtime suspend the DSP. That usually means switching the DSP off. Next time the
+guest wants to activate audio it will have to perform initialisation again,
+including sending its topology to the host. This happens automatically as long
```

Comment:
the topology file is provided by the host to the guest, I guess you must mean using IPC to configure the graph on the DSP during the topology parsing?

Reply:
@plbossart I mean booting the DSP for each runtime-resume. When the DSP boots, the driver has to re-send the topology to it. This is what is done now at least in some configurations. Here we have to do the same but we need to send all topology fragments - from the host and any guests.

Reply:
the host sends the topology binary to the guest.
The guest will issue a set of IPC to create the topology on the DSP.

Can you detail the steps on resume, e.g. can we assume the guest has kept a copy of the topology file and only needs to recreate the graph? or has the guest lost all context?